app/scripts/metamask-controller.js

```diff
@@ -4366,14 +4416,15 @@ export default class MetamaskController extends EventEmitter {
    * @returns {Promise<boolean>}
    */
   async forgetDevice(deviceName) {
-    const keyring = await this.getKeyringForDevice(deviceName);
+    return this.withKeyringForDevice({ name: deviceName }, async (keyring) => {
+      for (const address of keyring.accounts) {
+        await this._onAccountRemoved(address);
```

Comment:
By extracting side effects from `removeAccount` we can execute side effects here without ending up in a deadlock.

The account is removed from KeyringController anyway when calling `keyring.forgetDevice()`

Reply:
Hmm. These side effects are a little strange (permissions should be automatically removed when an account is deleted, for instance). But I see that this is similar to how it already works today. Seems like a reasonable solution to me

Reply:
I agree. Also, ideally, side effects should be executed _after_ the account removal instead of _before_ since an error could occur (that now would also trigger a rollback)